src/esp/gfx/DebugLineRender.cpp

```diff
@@ -0,0 +1,314 @@
+
+  // Update shader
+  _glResources->mesh.setCount(_verts.size());
+
+  glDisable(GL_LINE_SMOOTH);  // anti-aliased lines cause artifacts
+
+  glLineWidth(_internalLineWidth);
+
+  // We draw lines multiple times, with pixel offsets, to produce a single
+  // thick, visually-appealing line. Todo: implement thick lines using
+  // triangles, for example https://mattdesl.svbtle.com/drawing-lines-is-hard.
+  // 1.2 is hand-tuned for Nvidia hardware to be just small enough so we don't
+  // see gaps between the individual offset lines.
+  const float x = _internalLineWidth * 1.2f;
+  constexpr float sqrtOfTwo = 1.4142f;
+  // hard-coding 8 points around a circle
+  const std::vector<Mn::Vector3> offsets = {Mn::Vector3(x, x, 0),
+                                            Mn::Vector3(-x, x, 0),
+                                            Mn::Vector3(x, -x, 0),
+                                            Mn::Vector3(-x, -x, 0),
+                                            Mn::Vector3(x * sqrtOfTwo, 0, 0),
+                                            Mn::Vector3(-x * sqrtOfTwo, 0, 0),
+                                            Mn::Vector3(0, x * sqrtOfTwo, 0),
+                                            Mn::Vector3(0, -x * sqrtOfTwo, 0)};
+
+  Mn::Matrix4 projCam = projMatrix * camMatrix;
+
+  auto submitLinesWithOffsets = [&]() {
+    for (const auto& offset : offsets) {
+      // We want to offset in viewport (pixel) space, but we need to specify our
+      // offset transform in projection space. So divide by viewport dim.
+      Magnum::Matrix4 offset0Matrix = Mn::Matrix4::translation(
+          offset * Mn::Vector3(1.f / viewport.x(), 1.f / viewport.y(), 0.f));
+      Magnum::Matrix4 transProj = offset0Matrix * projCam;
+      _glResources->shader.setTransformationProjectionMatrix(transProj);
+      _glResources->shader.draw(_glResources->mesh);
+    }
+  };
+
+  submitLinesWithOffsets();
+
+  // modify all colors to be semi-transparent
+  // perf todo: do a custom shader constant for opacity instead so we don't have
+  // to touch all the verts
+  static float opacity = 0.1;
+  for (int v = 0; v < _verts.size(); v++) {
+    _verts[v].color.w() *= opacity;
+  }
+  _glResources->buffer.setData(_verts, Mn::GL::BufferUsage::DynamicDraw);
+
+  // Here, we re-draw lines with a reversed depth function. This causes
+  // occluded lines to be visualized as semi-transparent, which is useful for
+  // UX and visually-appealing.
+  Mn::GL::Renderer::setDepthFunction(Mn::GL::Renderer::DepthFunction::Greater);
+  submitLinesWithOffsets();
+
+  // restore to a reasonable default
+  Mn::GL::Renderer::setDepthFunction(Mn::GL::Renderer::DepthFunction::Less);
+
+  // Clear _verts to receive new data
+  arrayResize(_verts, 0);
+
+  // restore blending state if necessary
+  if (doToggleBlend) {
+    Mn::GL::Renderer::disable(Mn::GL::Renderer::Feature::Blending);
+    // Note: because we are disabling blending, we don't need to restore
```

Comment:
@mosra any concern here?

Reply:
Same as [in the other comment](https://github.com/facebookresearch/habitat-sim/pull/1349#discussion_r660108712), this should be something Magnum eventually takes care of. 

For now, this is fine.